Add Maven profile to automatically reformat code

kohlschuetter opened the pull request:
Jetty code can now automatically be reformatted by running the following command:

    mvn process-sources -Preformat

This will invoke the "reformat" Maven profile, which currently executes the "impsort-maven-plugin" to organize imports, and the "formatter-maven-plugin" to reformat Java source code, according to the rules defined in build/build-resources/jetty-codestyle-eclipse-ide.xml.

Note that not all code is currently formatted according to these rules, so a separate commit should incorporate the reformatting changes in the future.

Signed-off-by: Christian Kohlschütter <christian@kohlschutter.com>

kohlschuetter committed:
Add Maven profile to automatically reformat code

Jetty code can now automatically be reformatted by running the following
command:

    mvn process-sources -Preformat

This will invoke the "reformat" Maven profile, which currently executes
the "impsort-maven-plugin" to organize imports, and the
"formatter-maven-plugin" to reformat Java source code, according to the
rules defined in build/build-resources/jetty-codestyle-eclipse-ide.xml.

Note that not all code is currently formatted according to these rules,
so a separate commit should incorporate the reformatting changes in the
future.

https://github.com/eclipse/jetty.project/issues/9302

Signed-off-by: Christian Kohlschütter <christian@kohlschutter.com>

joakime reviewed (changes requested):
It would be better to base any kind of maven plugin that does reformatting off of the IntelliJ formatting ruleset, as it can actually satisfy the checkstyle ruleset.

If this codebase was stuck in Java 8, I'd say we can give this a try.
But we are using Java 17 now, which has many new language constructs that the Eclipse formatter has no concepts of and often gets wrong.

joakime commented:
Examples of the silliness of this plugin's results.

``` patch
-            .onRequestQueued(request -> { /* ... */ })
-            .onRequestBegin(request -> { /* ... */ })
-            .onRequestHeaders(request -> { /* ... */ })
-            .onRequestCommit(request -> { /* ... */ })
-            .onRequestContent((request, content) -> { /* ... */ })
-            .onRequestFailure((request, failure) -> { /* ... */ })
-            .onRequestSuccess(request -> { /* ... */ })
+            .onRequestQueued(request ->
+            {
+                /* ... */ })
+            .onRequestBegin(request ->
+            {
+                /* ... */ })
+            .onRequestHeaders(request ->
+            {
+                /* ... */ })
+            .onRequestCommit(request ->
+            {
+                /* ... */ })
+            .onRequestContent((request, content) ->
+            {
+                /* ... */ })
+            .onRequestFailure((request, failure) ->
+            {
+                /* ... */ })
+            .onRequestSuccess(request ->
+            {
+                /* ... */ })
```

and

``` patch
-        transport.setConnectionPoolFactory(destination ->
-            new RoundRobinConnectionPool(destination,
-                maxConnectionsPerDestination,
-                maxRequestsPerConnection));
+        transport.setConnectionPoolFactory(destination -> new RoundRobinConnectionPool(destination,
+            maxConnectionsPerDestination,
+            maxRequestsPerConnection));
```

and

``` patch
         Resource resource = ResourceFactory.combine(
             ResourceFactory.of(handler).newResource("/path/to/static/resources/"),
-            ResourceFactory.of(handler).newResource("/another/path/to/static/resources/")
-        );
+            ResourceFactory.of(handler).newResource("/another/path/to/static/resources/"));
```

and

``` patch
-        ServerSessionListener sessionListener = new ServerSessionListener() {};
+        ServerSessionListener sessionListener = new ServerSessionListener()
+        {
+        };
```

and

``` patch
-    private static final ChunkCallback EOF = new ChunkCallback(Content.Chunk.EOF, () -> {}, x -> {});
+    private static final ChunkCallback EOF = new ChunkCallback(Content.Chunk.EOF, () ->
+    {
+    }, x ->
+    {
+    });
```

and

``` patch
     @ParameterizedTest
-    @ValueSource(booleans = {true, false})
+    @ValueSource(booleans =
+    {true, false})
     public void testEncryptedOutputBufferRepoolingAfterNetworkFlushThrows(boolean close) throws Exception
```

and

``` patch
         return
-            // Partial right overlap: 10-20,15-30.
-            (range.first >= this.first && range.first <= this.last) ||
+        // Partial right overlap: 10-20,15-30.
+        (range.first >= this.first && range.first <= this.last) ||
             // Partial left overlap: 20-30,15-25.
             (range.last >= this.first && range.last <= this.last) ||
```

However, it does cleanup excessive whitespace usage nicely.

``` patch
-                for (int i = values.length; i-- > 0; )
+                for (int i = values.length; i-- > 0;)
```

and

``` patch
-        fields.put("path",  cookie.getPath());
+        fields.put("path", cookie.getPath());
```

and

``` patch
-    static final  Logger LOG = LoggerFactory.getLogger(MimeTypes.class);
+    static final Logger LOG = LoggerFactory.getLogger(MimeTypes.class);
```

joakime commented:
Running this plugin will modify 1,345 source files.

It will produce an environment that will fail checkstyle in the following places.

```
[WARN] /home/joakim/code/jetty/jetty.project-alt/jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/SerializedInvoker.java:210:86: '{' at column 86 should be on a new line. [LeftCurly]
[WARN] /home/joakim/code/jetty/jetty.project-alt/jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SniX509ExtendedKeyManager.java:135:44: ':' should be on a new line. [OperatorWrapNextLine]
[WARN] /home/joakim/code/jetty/jetty.project-alt/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/ByteRange.java:33:48: '{' at column 48 should be on a new line. [LeftCurly]
[WARN] /home/joakim/code/jetty/jetty.project-alt/jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/MultiPartCaptureTest.java:165:57: '{' at column 57 should be on a new line. [LeftCurly]
[WARN] /home/joakim/code/jetty/jetty.project-alt/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java:414:101: '{' at column 101 should be on a new line. [LeftCurly]
[WARN] /home/joakim/code/jetty/jetty.project-alt/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java:493:72: '{' at column 72 should be on a new line. [LeftCurly]
[WARN] /home/joakim/code/jetty/jetty.project-alt/jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/InputStreamResponseListener.java:365:112: '{' at column 112 should be on a new line. [LeftCurly]
[WARN] /home/joakim/code/jetty/jetty.project-alt/jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/ProxyProtocolClientConnectionFactory.java:607:54: ':' should be on a new line. [OperatorWrapNextLine]
[WARN] /home/joakim/code/jetty/jetty.project-alt/jetty-core/jetty-fcgi/jetty-fcgi-client/src/main/java/org/eclipse/jetty/fcgi/generator/Flusher.java:129:115: '{' at column 115 should be on a new line. [LeftCurly]
[WARN] /home/joakim/code/jetty/jetty.project-alt/jetty-core/jetty-fcgi/jetty-fcgi-client/src/main/java/org/eclipse/jetty/fcgi/parser/ClientParser.java:54:116: '{' at column 116 should be on a new line. [LeftCurly]
[WARN] /home/joakim/code/jetty/jetty.project-alt/jetty-core/jetty-http3/jetty-http3-common/src/main/java/org/eclipse/jetty/http3/ControlFlusher.java:164:58: '{' at column 58 should be on a new line. [LeftCurly]
[WARN] /home/joakim/code/jetty/jetty.project-alt/jetty-core/jetty-rewrite/src/test/java/org/eclipse/jetty/rewrite/handler/RewriteRegexRuleTest.java:143:124: '{' at column 124 should be on a new line. [LeftCurly]
[WARN] /home/joakim/code/jetty/jetty.project-alt/jetty-core/jetty-session/src/main/java/org/eclipse/jetty/session/AbstractSessionManager.java:1171:107: '{' at column 107 should be on a new line. [LeftCurly]
[WARN] /home/joakim/code/jetty/jetty.project-alt/jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/HttpClientStreamTest.java:1164:115: '{' at column 115 should be on a new line. [LeftCurly]
[WARN] /home/joakim/code/jetty/jetty.project-alt/jetty-core/jetty-websocket/jetty-websocket-core-server/src/main/java/org/eclipse/jetty/websocket/core/server/internal/WebSocketNegotiation.java:121:74: ':' should be on a new line. [OperatorWrapNextLine]
[WARN] /home/joakim/code/jetty/jetty.project-alt/documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/HTTP2Docs.java:53:64: '{' at column 64 should be on a new line. [LeftCurly]
```

kohlschuetter committed:
Sort jetty-codestyle-eclipse-ide.xml

... so we can better compare changes

kohlschuetter committed:
Import/Re-export jetty-codestyle-eclipse-ide.xml to Eclipse 2022-12

This will add missing formatting configurations that are necessary for
formatter-maven-plugin to work correctly.

kohlschuetter committed:
Fix jetty-codestyle-eclipse-ide.xml to not fail checkstyle:check

The Eclipse formatter (both in Eclipse IDE and using the Maven plugin)
would reformat some code in a way that would fail checkstyle:check.

Fix conditional wrapping, and also keep annotation declarations one one
line more often than not.

kohlschuetter commented:
> Examples of the silliness of this plugin's results.

I wouldn't call this "silliness". I guess you would just need to tweak the configuration a little more to your needs.

After all, Jetty's rules are not simple.

Some of your examples claim that putting an opening brace on a new line is bad, but that's the default for most other cases in this project (method bodies, etc.).

I bet all these things are configurable. I will poke a little more on this purely out of my own interest. However, with the latest change, the formatter passes `checkstyle:check`, so ... ?

kohlschuetter committed:
More tweaks to jetty-codestyle-eclipse-ide.xml

joakime commented:
This update would now impact 1,319 source files.

It is still forcing new line on open brace everywhere, not what we want.
Esp if there are no statements inside of the braces.

Example:

``` patch
             // Add request hooks.
-            .onRequestQueued(request -> { /* ... */ })
-            .onRequestBegin(request -> { /* ... */ })
-            .onRequestHeaders(request -> { /* ... */ })
-            .onRequestCommit(request -> { /* ... */ })
-            .onRequestContent((request, content) -> { /* ... */ })
-            .onRequestFailure((request, failure) -> { /* ... */ })
-            .onRequestSuccess(request -> { /* ... */ })
+            .onRequestQueued(request ->
+            {
+                /* ... */ })
+            .onRequestBegin(request ->
+            {
+                /* ... */ })
+            .onRequestHeaders(request ->
+            {
+                /* ... */ })
+            .onRequestCommit(request ->
+            {
+                /* ... */ })
+            .onRequestContent((request, content) ->
+            {
+                /* ... */ })
+            .onRequestFailure((request, failure) ->
+            {
+                /* ... */ })
+            .onRequestSuccess(request ->
+            {
+                /* ... */ })
```

and

``` patch
-        CompletableFuture<Session> sessionCF = http2Client.connect(serverAddress, new Session.Listener() {});
+        CompletableFuture<Session> sessionCF = http2Client.connect(serverAddress, new Session.Listener()
+        {
+        });
```

The intellij formatter ruleset and checkstyle honors these.
Can the eclipse formatter do this?

There is also a collapsing of end parentheses onto the same line for multi-line formatted statements.

Example:

``` patch
         // For multiple directories, use ResourceCollection.
         Resource resource = ResourceFactory.combine(
             ResourceFactory.of(handler).newResource("/path/to/static/resources/"),
-            ResourceFactory.of(handler).newResource("/another/path/to/static/resources/")
-        );
+            ResourceFactory.of(handler).newResource("/another/path/to/static/resources/"));
```

This is also supported by both the intellij formatter and checkstyle.
We don't want these kinds of changes.

There is also a strange behavior (in both formatters, and the checkstyle rules) when multiple lambdas are used as parameters on a method (or constructor)

``` patch
-    private static final ChunkCallback EOF = new ChunkCallback(Content.Chunk.EOF, () -> {}, x -> {});
+    private static final ChunkCallback EOF = new ChunkCallback(Content.Chunk.EOF, () ->
+    {
+    }, x ->
+    {
+    });
```

The checkstyle ruleset should allow the first mode, but if the second mode is used, then it should have put the extra param on a new line (looking like this) ...

``` java
    private static final ChunkCallback EOF = new ChunkCallback(Content.Chunk.EOF, () ->
    {
    }, 
    x ->
    {
    });
```

I would have expected a checkstyle failure on the altered line. Have to investigate this quirk.

The changes to multi-line ternary operations are also odd.

``` patch
-                int vectorsLength = tlvs == null ? 0 : tlvs.stream()
-                    .mapToInt(tlv -> 1 + 2 + tlv.getValue().length)
-                    .sum();
+                int vectorsLength = tlvs == null ? 0
+                    : tlvs.stream()
+                        .mapToInt(tlv -> 1 + 2 + tlv.getValue().length)
+                        .sum();
                 capacity += vect
```

The forced new line for `:` is not ideal.
The formatter should not have modified this line (intellij formatter supports this)

Also a weird indentation change on comments in array parameters.

``` patch
             return Stream.of(
                 new NormalScenario(),
                 new SslScenario()
-                // TODO: add more ssl / non-ssl scenarios here
+            // TODO: add more ssl / non-ssl scenarios here
             ).map(Arguments::of);
```

I would have expected that to have no indentation change.
The formatter should not have modified this line (intellij formatter supports this)

The removal of new lines is also undesired.

``` patch
-        ExecutionException x = assertThrows(ExecutionException.class, () ->
-            client.newRequest("localhost", connector.getLocalPort())
-                .scheme(scenario.getScheme())
-                .method(HttpMethod.DELETE)
-                .path("/301/localhost/done")
-                .timeout(5, TimeUnit.SECONDS)
-                .send());
+        ExecutionException x = assertThrows(ExecutionException.class, () -> client.newRequest("localhost", connector.getLocalPort())
+            .scheme(scenario.getScheme())
+            .method(HttpMethod.DELETE)
+            .path("/301/localhost/done")
+            .timeout(5, TimeUnit.SECONDS)
+            .send());
```

This example makes it difficult to see the purpose/role of the lambda.
The formatter should not have modified this line (intellij formatter supports this)

The forced newline on open brace is undesired in some contexts.

``` patch
-    @ValueSource(booleans = {true, false})
+    @ValueSource(booleans =
+    {true, false})
```

The formatter should not have modified this line (intellij formatter supports this)

**Now for the good stuff!**

The indentation fixes, comment alignment fixes, excessive whitespace fixes, are all good.

joakime commented:
Just to be clear, I'm not the only voice in this project. :wink: 

Just the first one to chime in.
I hope that others will speak up about the discoveries so far.

Meanwhile, this analysis has revealed 2 quirks in the checkstyle rules I want to address (in a different PR)

1. The odd behaviors when using lambdas in a multi parameter method/constructor.
2. The new "record" type and brace handling.

joakime commented on build/build-resources/reformat-modified:
```diff
@@ -0,0 +1,31 @@
+      echo "$f,"
+  fi 
+done)
+$(git ls-files --others --exclude-standard | grep \.java )
```
This should only modify changed lines and import section, no other lines in the source file should be altered.

(the Intellij formatter supports this)

kohlschuetter replied:
This is as much as I'm able to contribute at this point.

I'm really trying to help here, but I feel you set the bar too high, @joakime.
I'm not going to rewrite the Eclipse formatter for you.

kohlschuetter committed:
Add "reformat-modified" script to only reformat changed/new files

This reduces the scope of arguing about code formatting subtleties to
the set of modified and/or added files.

Signed-off-by: Christian Kohlschütter <christian@kohlschutter.com>

kohlschuetter commented:
With the recent changes, please re-evaluate the output of
`mvn process-sources -Preformat` 

Also check out the script `build/build-resources/reformat-modified`, which reformats only the modified/new files according to `git diff`.

I'm leaving this PR as-is for now. It's good enough™ for my needs. Feel free to improve/merge/abandon as you wish. I've already spent too much time bikeshedding.

janbartel commented:
@kohlschuetter I clearly didn't understand the workflow of this proposal when I first commented. I get it now :)

I am the aforementioned Eclipse IDE user and the one thing that I really struggle with is satisfying jetty's `import` order rules. There seems to be no way to configure Eclipse to generate what is required, so I always have to format those by hand. I guess the reformatter plugin will have the same difficulty?

olamy commented:
as we are in the proposal/discussion stage, I would go in a different direction ;) 
currently we have to maintain 3 files with a totally different format and it's a sort of pain to have run a build or `mvn checkstyle:check` and then return to your IDE to fix failing files.
What about using a totally different strategy and using only 1 configuration file and a single maven plugin..
At the end of the day, do we care really care about IDE if you only need to run something such `mvn spotless:appy` and this will format java files and pom files. Because we do have some style differences between poms which can be/is a pain when merging 11 to 12...
The normal build should use `mvn spotless:check` rather than `checkstyle:check` and when you need to format code because you did some changes just run `mvn spotless:apply`.
So we only have to focus on code changes and don't have to struggle coming back and forward to IDE too much with the style because a tool will do it automatically for us.

The problem now is to find the correct spotless configuration which can fit our current code style configuration to not change anything regarding styles (except pom formatting will change few)
this plugin can do both check/reformat for java sources and pom https://github.com/diffplug/spotless/tree/main/plugin-maven
maybe we can find others doing similar?

kohlschuetter commented:
@janbartel The "reformat" profile runs two plugins, one that is responsible for code formatting, the other one for organizing imports. I've tuned both well enough so that they pass jetty checkstyle. The organize-imports part is pretty robust, I've used it in https://github.com/eclipse/jetty.project/pull/9300, which just got merged. With this change, I hope you can keep using Eclipse IDE, that's what I'm using, too.

@olamy [XKCD 927](https://xkcd.com/927/) comes to mind. I haven't used spotless myself, but it looks like it may actually reach out to the Eclipse JDT formatter under the hood (unless you want Google or Palantir formatting).

From my own experience with checkstyle, I would simply stop using it for verifying code _formatting_ (checkstyle should still be used to find bugs, not complain about the wrong number of whitespace), and indeed rely on a single source of truth for formatting instead (a commonly agreed upon set of formatting rules that can be enforced from the command line).

As far as I understand, the Eclipse formatting rules are the closest to achieving this:

1. They can be run via Maven
2. They can obviously be run via Eclipse IDE
3. They can (AFAIK) be imported into IntelliJ

joakime commented:
Not going to use this.  Closing.
